File: problems/p003/solution.java
/**
 * Project Euler: Problem #003
 *
 * @author   Alexander Arzhanov
 * @version  07/02/18
 */
 
package problems.p003;
import problems.util.EulerSolvable;


// Largest prime factor
public final class p003 implements EulerSolvable {

    /**
    * Solve by factorizing the input parameter.
    *
    * @param in  Number to factorize (String in in[0]).
    * @return    The largest prime factor (String).
    */
    @Override
    public String solve(String... in) {
        
        long num = Long.parseLong(in[0]);
        long ans = 0;
        
        // case: 0 or 1
        if (num < 2) return Long.toString(num);
        
        // reduce even
        while (num % 2 == 0) {
            num /= 2;
            ans = 2;
        }
        
        // reduce odd
        long lim = (long) Math.sqrt(num) + 1;
        long i = 3;
        
        while (i <= lim) {
            
            if (num % i == 0) {
                num /= i;
                lim = (long) Math.sqrt(num) + 1;
                ans = i;
                System.out.println(ans);
                continue;
            }
            
            i += 2;
        }
        
        // add the remainder factor (also self-prime case)
        if (num != 1) ans = num;
        
        return Long.toString(ans);
    }
    
    // Stand-alone test-run
    public static void main(String args[]) {
        
        System.out.println("\n ANS: " + new p003().solve("13195"));
        
    }    
}

File: problems/p003/solution.py
"""
Project Euler: Problem #003
"""

from math import sqrt

def solve(num):
    """
    Solve by factorizing the input parameter.

    :param num: Number to factorize.
    :returns: The largest prime factor.
    """
    
    # check validity of input
    assert(num >= 0)
    
    # check 0 & 1
    if num < 2: 
        return num
        
    res = []
    
    # reduce even
    while num % 2 == 0:
        res.append(2)
        num //= 2
        
    # reduce odd
    lim = sqrt(num)
    i = 3
    
    while i <= lim:
    
        if num % i == 0:
            res.append(i)
            num //= i
            lim = sqrt(num)
            
        else:
            i += 2
            
    if num != 1:
        res.append(num)
        
    return max(res)
        